python/ray/llm/_internal/batch/stages/sglang_engine_stage.py

```diff
@@ -0,0 +1,367 @@
+"""The stage that runs SGLang engine."""
+
+import asyncio
+import logging
+import time
+import uuid
+from enum import Enum
+from pydantic import BaseModel, root_validator
+from typing import Any, Dict, AsyncIterator, Optional, List, Tuple, Type
+
+from ray.llm._internal.utils import try_import
+from ray.llm._internal.batch.stages.base import (
+    StatefulStage,
+    StatefulStageUDF,
+)
+
+sgl = try_import("sglang")
+
+logger = logging.getLogger(__name__)
+
+
+class SGLangTaskType(str, Enum):
+    """The type of task to run on the SGLang engine."""
+
+    """Generate text."""
+    GENERATE = "generate"
+
+    # TODO: Add support for the following tasks.
+    """Generate embeddings."""
+    EMBED = "embed"
+
+    """Classify the quality of pairwise generations."""
+    REWARD = "reward"
+
+
+class SGLangEngineRequest(BaseModel):
+    """A request to the SGLang engine."""
+
+    # The request ID for the LLM engine (unique per replica).
+    request_id: int
+    # The index of the request in the batch.
+    idx_in_batch: int
+    # The input prompt.
+    text: Optional[str]
+    # Alternative to text. Specify the input as token IDs instead of text.
+    input_ids: Optional[List[int]]
+    # The sampling parameters (more details can be seen in https://docs.sglang.ai/backend/sampling_params.html).
+    sampling_params: Optional[Dict[str, Any]]
+
+    class Config:
+        validate_assignment = True
+        arbitrary_types_allowed = True
+
+
+class SGLangOutputData(BaseModel):
+    """The output of the SGLang engine."""
+
+    prompt: Optional[str]
+    prompt_token_ids: Optional[List[int]]
+    num_input_tokens: int
+
+    # Generate fields.
+    generated_tokens: Optional[List[int]]
+    generated_text: Optional[str]
+    num_generated_tokens: int
+
+    # Metrics fields.
+    metrics: Optional[Dict[str, Any]] = None
+
+    @classmethod
+    def from_sglang_engine_output(cls, output: Dict[str, Any]) -> "SGLangOutputData":
+        """Create a SGLangOutputData from a SGLang engine output."""
+
+        data = cls(
+            prompt=output["prompt"],
+            prompt_token_ids=output["prompt_token_ids"],
+            num_input_tokens=output["meta_info"]["prompt_tokens"],
+            generated_tokens=output["output_ids"] if "output_ids" in output else None,
+            generated_text=output["text"] if "text" in output else None,
+            num_generated_tokens=output["meta_info"]["completion_tokens"],
```

Comment:
same question here.

Reply:
Yes. Refer to the output example in https://docs.sglang.ai/backend/native_api.html#Generate-(text-generation-model).